sci-physics/geant-vmc: new package.

olifre opened the pull request:
Geant4 VMC represents the realisation of the
Virtual Monte Carlo (VMC) for Geant4.

Package-Manager: Portage-2.3.24, Repoman-2.3.6
RepoMan-Options: --force

Package has lived in Science Overlay for a few years, I have bumped and polished it a bit. 
Please let me know about any remaining issues and I'll address them quickly. 

**Note: This package depends on #6922 which has been stalled in review for a few weeks.**

doughdemon commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,76 @@
+
+src_compile() {
+	cmake-utils_src_compile
+	local dirs="source"
```
Starting from here, make everything dependent on `use doc` instead checking `use doc` every time.

doughdemon commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,76 @@
+src_compile() {
+	cmake-utils_src_compile
+	local dirs="source"
+	use g4root && dirs+=" g4root "
```
By using `usev` you can avoid some repetition. (It might also be nicer to make `dirs` a bash array.)

olifre commented:
Thanks for the very helpful recommendations!
I have:
- Changed over to use of a bash array, and now use `usev`.
- Check `use doc` only once.
- Fixed `mtroot` use flag, apparently I missed upstream renaming the cmake flag. 
- Added a live ebuild.

monsieurp commented on sci-physics/geant-vmc/geant-vmc-9999.ebuild:
```diff
@@ -0,0 +1,78 @@
+HOMEPAGE="http://root.cern.ch/root/vmc/VirtualMC.html"
+
+LICENSE="GPL-2"
+SLOT="4"
```
This is a 9999 ebuild. Please remove the `SLOT`.

olifre replied:
The intent of the SLOT operator was to differentiate this live ebuild from the Geant3-VMC code, which pairs with sci-physics/geant:3 (not packaged anymore...). 

If geant:3 was still alive, also `=geant-vmc-9999:3` might be a packaged version which could coexist with `=geant-vmc-9999:4` (both would build from a separate repository and depend on matching `sci-physics/geant` SLOTs). 

So it is more like e.g. `=sys-libs/compiler-rt-sanitizers-9999:7.0.0` which currently lives in tree. 

Should the SLOT still be removed? 
Note that
```
EGIT_REPO_URI="https://github.com/vmc-project/geant4_vmc.git"
```
and a future `GeantV` version of VMC (no idea if this will be made) would for sure be in a separate repository again, and could coexist for a while.

olifre commented:
Squashed, but take note of my question about the `SLOT`.

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+
+EAPI=6
+
+inherit cmake-utils versionator
```
Please use `eapi7-ver` instead of `versionator`.

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+
+if [[ ${PV} == *9999* ]]; then
+	inherit git-r3
+	EGIT_REPO_URI=" http://root.cern.ch/git/geant4_vmc.git"
```
Weird-ish space before the URL.

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+HOMEPAGE="http://root.cern.ch/root/vmc/VirtualMC.html"
+
+LICENSE="GPL-2"
+SLOT="4"
```
Given this is a new package, any reason to use non-zero slot? Do you expect to have future versions that will be co-installable in parallel?

olifre replied:
@mgorny : This package is directly bound to `sci-physics/geant` and its `SLOT`s. 
Right now, there is only `sci-physics/geant:4` in tree (still the slot was kept).

In the past, there also was `sci-physics/geant:3` in-tree and a co-installable `sci-physics/geant-vmc:3` was living in the gentoo-science-overlay. 

Of course, it's hard to tell the future, but it is likely that GeantV ( http://geant.cern.ch/ ), whenever it sees some stable release will need a separate VMC interfacing library. 

I have no very strong opinion, but currently I'd tend to stay close to the `SLOT` choice also employed for `sci-physics/geant`. What do you think?

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+
+RDEPEND="
+	sci-physics/root:=
+	>=sci-physics/geant-4.10.03:=[opengl,geant3?]
```
Please order unconditional deps lexically by package name, this makes it easier to find stuff when there's a lot of it (yeah, I get this is not the case here).

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+
+src_configure() {
+	local mycmakeargs=(
+			-DGeant4VMC_USE_VGM="$(usex vgm)"
```
Please indent with a single extra tab, not two.

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+	cmake-utils_src_compile
+	if use doc ; then
+		local dirs=(
+				source
```
Likewise (dedent).

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+			)
+		local d
+		for d in "${dirs[@]}"; do
+			pushd ${d} > /dev/null || die
```
Please quote path variables whenever possible, especially that you quoted it above.

mgorny commented on sci-physics/geant-vmc/geant-vmc-4.3.6.ebuild:
```diff
@@ -0,0 +1,78 @@
+
+src_install() {
+	cmake-utils_src_install
+	use doc && local HTML_DOCS=(doc/*)
```
Please use `doc/.` to avoid unnecessary bash expansion.

mgorny commented on sci-physics/geant-vmc/metadata.xml:
```diff
@@ -0,0 +1,22 @@
+  <maintainer type="person">
+    <email>o.freyermuth@googlemail.com</email>
+    <name>Oliver Freyermuth</name>
+    <description>Proxy-maintainer, assign bugs</description>
```
Please remove this redundant description, bug assignment rules already say to assign to the first person.

olifre commented:
@mgorny  Thanks for the brilliantly detailed and very instructive review! 
I learned a lot, especially `eapi7-ver` was news to me. 
All changes apart from the `SLOT` change are in, see my earlier comment on this. 
Many thanks!

mgorny commented:
@amadio, if you have ROOT built, could you test this, please?

olifre commented:
This is indeed very peculiar. I will look into it, but I can already say that indeed I also see test failures if I check now, although the installed geant-vmc library works perfectly fine when used e.g. in our analysis application.

olifre commented:
@mgorny Looks like a good old sandbox issue. `sci-physics/root` is special, they write to `/dev/random` during execution, misusing that to have the kernel validate userspace pointers. Sandbox prevents that. 
I'll look if a sandbox exception helps.

olifre commented:
@mgorny It does, tests run through perfectly fine now on my side. 
Can you confirm?

mgorny commented:
Well, I can confirm that something had changed.

```
... 3 tests passed successfully.
... 25 tests failed.
```

All of them seem to segfault on me. However, I don't know if it's problem with this specific package or one of its dependencies.

olifre commented:
@mgorny That's strange, after the sandbox `add_write` added, they all work perfectly well here. 
Just to make sure, which version of `sci-physics/root` are you using? I only tried with `6.12.04-r2` up to now, since `5.34.36` is rather old (that old version also should not need the sandbox exception). 
Upstream claims compatibility with both versions, though. 

Can you upload one of the test logs, such as:
```
/var/tmp/portage/sci-physics/geant-vmc-4.3.6/work/geant4_vmc-3-6/examples/log/E01/test_g4_tgeo_nat.out
```
which should contain more information about what exactly segfaulted?

mgorny commented:
[test_g4_tgeo_nat.out.txt](https://github.com/gentoo/gentoo/files/1755376/test_g4_tgeo_nat.out.txt)

olifre commented:
@mgorny Thanks!
```
cling::DynamicLibraryManager::loadLibrary(): /usr/lib64/libG4particles.so: cannot allocate memory in static TLS block
```
looks like ROOT's cling can not load `libG4particles` dynamically via `dlopen()`. 
I've never observed this before, likely also:
```
$ root   
   ------------------------------------------------------------
  | Welcome to ROOT 6.12/04                http://root.cern.ch |
  |                               (c) 1995-2017, The ROOT Team |
  | Built for linuxx8664gcc                                    |
  | From tag v6-12-04, 13 December 2017                        |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q' |
   ------------------------------------------------------------

root [0] gSystem->Load("/usr/lib64/libG4particles.so")

```
will fail, which means it would be unrelated to this package (but shows up in this package's tests). 

This makes me believe your `sci-physics/root` and `sci-physics/geant:4` are incompatible. 
- What are the exact versions?
- Are they both compiled with the exact same version of `gcc`?
- Which version of `gcc` do you use?
- Any special `CFLAGS` / `CXXFLAGS`?
- There are known issues with unstable `jemalloc`, see e.g. https://github.com/commercialhaskell/stack/issues/3220 . Which version does your system have?

mgorny commented:
> What are the exact versions?

```
[ebuild   R   ~] sci-physics/root-6.12.04-r2:6.12/04::gentoo  USE="X asimage examples fftw fortran gdml gsl math minuit opengl python roofit ssl tbb threads tiff tmva unuran xml -R (-aqua) -avahi -davix -emacs -fits -graphviz -http -jemalloc -kerberos -ldap -libcxx -memstat -mysql -odbc -oracle -postgres (-prefix) -pythia6 -pythia8 (-qt4) -root7 -shadow -sqlite -table -test -vc -xinetd -xrootd" PYTHON_SINGLE_TARGET="python3_5 -python2_7 -python3_4 -python3_6" PYTHON_TARGETS="python2_7 python3_4 python3_5 python3_6" 0 KiB
[ebuild   R   ~] sci-physics/geant-4.10.03-r1:4::gentoo  USE="data opengl qt5 threads zlib -dawn -doc -examples -gdml -geant3 -inventor -motif -raytracerx -static-libs -vrml" 0 KiB
```

> Are they both compiled with the exact same version of gcc?

Yep, built them yesterday-today specifically to test this.

> Which version of gcc do you use?

```
[ebuild   R    ] sys-devel/gcc-6.4.0-r1:6.4.0::gentoo  USE="cxx fortran graphite (multilib) nls nptl objc objc++ objc-gc openmp pch (pie) sanitize ssp vtv (-altivec) (-awt) -cilk -debug -doc (-fixed-point) (-gcj) -go (-hardened) (-jit) (-libssp) -mpx -pgo -regression-test -vanilla" 0 KiB
```

(that's 17.0+ profile)

> Any special CFLAGS / CXXFLAGS?

```
$ portageq envvar CFLAGS CXXFLAGS
-march=k8-sse3 -mcx16 -msahf --param l1-cache-size=64 --param l1-cache-line-size=64 --param l2-cache-size=512 -O2 -pipe -frecord-gcc-switches
-march=k8-sse3 -mcx16 -msahf --param l1-cache-size=64 --param l1-cache-line-size=64 --param l2-cache-size=512 -O2 -pipe -frecord-gcc-switches
```

> There are known issues with unstable jemalloc, see e.g. commercialhaskell/stack#3220 . Which version does your system have?

None ;-).

olifre commented:
@mgorny Thanks! That's strange, I have `jemalloc` here, but then it should not cause the issue. 
Could you (as normal user) just run:
`root`
and then type
```
gSystem->Load("/usr/lib64/libG4particles.so")
```
on the interpreter?
If this fails, at least we know it's really an incompatibility between the two. 

I have exactly the same versions you have, but different flags:
```
[ebuild   R   ~] sci-physics/root-6.12.04-r2:6.12/04::gentoo  USE="X asimage avahi davix emacs examples fftw fortran gdml graphviz gsl http kerberos ldap math memstat minuit mysql odbc opengl postgres pythia6 python roofit shadow sqlite ssl tbb threads tiff tmva unuran vc xml xrootd -R (-aqua) -fits -jemalloc -libcxx -oracle (-prefix) -pythia8 -qt4 -root7 -table {-test} -xinetd" PYTHON_SINGLE_TARGET="python3_5 -python2_7 -python3_4 -python3_6" PYTHON_TARGETS="python2_7 python3_5 -python3_4 -python3_6" 0 KiB
[ebuild   R   ~] sci-physics/geant-4.10.03-r1:4::gentoo  USE="data dawn gdml geant3 motif opengl qt5 raytracerx vrml zlib -doc -examples -inventor -static-libs -threads" 0 KiB
[ebuild   R    ] sys-devel/gcc-6.4.0-r1:6.4.0::gentoo  USE="cilk cxx fortran go graphite mpx (multilib) nls nptl objc objc++ openmp pch pgo (pie) sanitize ssp vtv (-altivec) (-awt) -debug -doc (-fixed-point) (-gcj) (-hardened) (-jit) (-libssp) -objc-gc -regression-test -vanilla" 0 KiB
```
also on a 17.0 profile, with:
```
$ portageq envvar  CFLAGS CXXFLAGS
-march=haswell -mabm -O2 -pipe
-march=haswell -mabm -O2 -pipe
```

mgorny commented:
```
root [1] gSystem->Load("/usr/lib64/libG4particles.so")
cling::DynamicLibraryManager::loadLibrary(): /usr/lib64/libG4particles.so: cannot allocate memory in static TLS block
(int) -1
```

olifre commented:
Thanks! This confirms ROOT can not load Geant's libraries on your system as it stands. 

A notable difference is that I am using:
```
sci-physics/geant -threads
```
Could you test if things work after disabling this flag? 
Alternatively, I can also test whether things break when I set the flag, just let me know.

olifre commented:
@mgorny Could you please also post the output of:
```
$ geant4-config --cflags
```
before recompiling with `-threads`?
I expect this to contain:
```
-ftls-model=initial-exec -pthread
```
which would explain why it's shared libraries are incompatible with anything else on the system which has another TLS model.

mgorny commented:
Yep, it has that.

olifre commented:
This explains it, then. 
I should depend on `sci-physics/geant[-threads]` explicitly, then, but in the end, this also means the libraries installed by Geant4 with `threads` can not be used with anything else on the system, unless everything in the dependency chain is compiled with this TLS model.
I'm unsure whether that warrants a bug against `sci-physics/geant`, this TLS model appears to have been upstream's decision. The manpage of gcc tells us:
```
The default without -fpic is initial-exec; with -fpic the default is global-dynamic.
```

mgorny commented:
Please do report a bug, and let the Gentoo maintainer decide how to handle this.

olifre commented:
Done:
https://bugs.gentoo.org/648736
So now there are two ways to proceed:
- Depend on `sci-physics/geant[-threads]` explicitly with this package. 
- Wait for decision on bug 648736 and decide afterwards. 
Since I am in no hurry, I would be fine with both.

amadio commented:
Hi @mgorny, @olifre. I will talk to Gabriele to see if we can just remove the tls option from Geant4, then I will test and merge this. Thanks for the contribution!

olifre commented:
@amadio Perfect, many thanks! Did you also get a reply from upstream (Gabriele) whether this default flag may be dropped in general in a future upstream release? Or is there a real existing need for it?

amadio commented:
@olifre Yes, I don't think this will change in Geant4, but there is an option to set the threading model that we can adjust in the ebuild.

olifre commented:
I can confirm that after recompiling `=sci-physics/geant-4.10.03-r1[thread]` as of 395a5469e0d79b72275ebb16275b1957dc65a7f4 now all tests of `sci-physics/geant-vmc` run through fine for me even for this "threaded" geant. Thanks! 

If that's the only open issue, I think it could be merged - I could of course also bump to `EAPI=7` and remove `eapi7-ver` if `cmake-utils` is already EAPI7-ready (did not check yet). 

Also, I guess `=sci-physics/geant-4.10.03-r1` needs a revbump to ensure everybody has the version with changed TLS model? I commented on https://bugs.gentoo.org/648736 on this matter.

mgorny commented:
Do I understand correctly that we can merge it now? If not, please shout before I finish re-testing ;-).

olifre commented:
@mgorny as of https://bugs.gentoo.org/648736 , @amadio first wanted to bump Geant to `=sci-physics/geant-4.10.04` to reduce the number of rebuilds (and ensure everybody has the changed `tls-model` flag). But apart from that, I think there's nothing blocking things ;-).

mgorny commented:
For the record, I've run out of space while testing this, so I won't be merging it today.

amadio commented:
Hi @olifre, I see errors with the tests:
```
>>> Source compiled.
>>> Test phase: sci-physics/geant-vmc-4.3.6
... Example E01
... Running test with G4, geometry via TGeo, Native navigation
... Running test with G4, geometry via TGeo, TGeo navigation
... Running test with G4, geometry via VMC,  Native navigation
... Running test with G4, geometry via VMC,  TGeo navigation
 
... Example E02
... Running test with G4, geometry via TGeo, Native navigation
./test_suite.sh: line 70: 21710 Aborted                 (core dumped) $RUNG4 "test_$EXAMPLE.C(\"g4Config.C\",kFALSE)" 1>&$OUT/test_g4_tgeo_nat.out
... Running test with G4, geometry via TGeo, TGeo navigation
./test_suite.sh: line 70: 21844 Aborted                 (core dumped) $RUNG4 "test_$EXAMPLE.C(\"g4tgeoConfig.C\",kFALSE)" 1>&$OUT/test_g4_tgeo_tgeo.out
... Running test with G4, geometry via VMC,  Native navigation
./test_suite.sh: line 70: 21990 Aborted                 (core dumped) $RUNG4 "test_$EXAMPLE.C(\"g4ConfigOld.C\",kTRUE)" 1>&$OUT/test_g4_vmc_nat.out
... Running test with G4, geometry via VMC,  TGeo navigation
./test_suite.sh: line 70: 22135 Aborted                 (core dumped) $RUNG4 "test_$EXAMPLE.C(\"g4tgeoConfigOld.C\",kTRUE)" 1>&$OUT/test_g4_vmc_tgeo.out
 
... Example E03
... Running test with G4, geometry via TGeo, Native navigation
./test_suite.sh: line 70: 22280 Aborted                 (core dumped) $RUNG4 "test_E03_1.C(\"g4Config.C\", kFALSE)" 1>&$OUT/test_g4_tgeo_nat.out
./test_suite.sh: line 70: 22424 Aborted                 (core dumped) $RUNG4 "test_E03_2.C(\"g4Config.C\", kFALSE)" &> tmpfile
./test_suite.sh: line 70: 22568 Segmentation fault      (core dumped) $RUNG4 "test_E03_3.C(\"g4Config.C\", kFALSE)" &> tmpfile
./test_suite.sh: line 70: 22727 Aborted                 (core dumped) $RUNG4 "test_E03_4.C(\"g4Config.C\", kFALSE)" &> tmpfile
./test_suite.sh: line 70: 22875 Aborted                 (core dumped) $RUNG4 "test_E03_5.C(\"g4Config4.C\", kFALSE)" &> tmpfile
./test_suite.sh: line 70: 23024 Aborted                 (core dumped) $RUNG4 "test_E03_6.C(\"g4Config5.C\", kFALSE)" &> tmpfile
... Running test with G4, geometry via TGeo, (gets stuck here)
```

amadio commented:
Here's some relevant information about what I have installed:
```
epsft-53 ~ # equery u geant root geant-vmc
[ Legend : U - final flag setting for installation]
[        : I - package is installed with flag     ]
[ Colors : set, unset                             ]
 * Found these USE flags for sci-physics/geant-4.10.03-r1:
 U I
 + + data        : Add a lot of standard physics data files for geant4
 + + dawn        : Add support for media-gfx/dawn (3D postscript rendering)
 + + doc         : Add extra documentation (API, Javadoc, etc). It is recommended to enable per package instead of globally
 + + examples    : Install examples, usually source code
 + + gdml        : Enable geometry markup language for xml
 + + geant3      : Add compatibility for geant321 to geant4
 + + inventor    : Add support for Open Inventor SGI toolkit via Coin3D
 + + motif       : Add support for the Motif toolkit
 + + opengl      : Add support for OpenGL (3D graphics)
 + + qt5         : Add support for the Qt 5 application and UI framework
 + + raytracerx  : Enable raytracing for physics events
 + + static-libs : Build static versions of dynamic libraries as well
 + + threads     : Add threads support for various packages. Usually pthreads
 + + vrml        : Enable output of geant4 in vrml formats
 + + zlib        : Add support for zlib (de)compression

 * Found these USE flags for sci-physics/root-6.13.02:
 U I
 + + R                              : Enable support for dev-lang/R
 + + X                              : Add support for X11
 + - asimage                        : Enable support for media-libs/libafterimage
 + - avahi                          : Add avahi/Zeroconf support
 + - davix                          : Enable support for DAVIX Data Management Client (https://cern.ch/davix)
 + - emacs                          : Add support for GNU Emacs
 + - examples                       : Install examples, usually source code
 + + fftw                           : Use FFTW library for computing Fourier transforms
 + - fits                           : Enable support for images and data from FITS files with sci-libs/cfitsio
 + + fortran                        : Compile included Fortran programs
 + - gdml                           : Enable GDML writer and reader
 + + graphviz                       : Add support for the Graphviz library
 + - gsl                            : Use the GNU scientific library for calculations
 + - http                           : Enable http server support including but not limited to fastcgi support
 + - jemalloc                       : Enable support for dev-libs/jemalloc
 + + kerberos                       : Add kerberos support
 + + ldap                           : Add LDAP support (Lightweight Directory Access Protocol)
 - - libcxx                         : Build ROOT against sys-libs/libcxx
 + - math                           : Build all math related libraries plugins, needs sci-libs/gsl 
 + - memstat                        : Build memory statistics library, helps to detect memory leaks
 + - minuit                         : Build ROOT's internal versions of minuit (fortran) and minuit2 (C++)
 + - mysql                          : Add mySQL Database support
 + - odbc                           : Add ODBC Support (Open DataBase Connectivity)
 + + opengl                         : Add support for OpenGL (3D graphics)
 + - oracle                         : Enable Oracle Database support
 + - postgres                       : Add support for the postgresql database
 + - pythia6                        : Build the interface for sci-physics/pythia version 6.x 
 + - pythia8                        : Build the interface for sci-physics/pythia version 8.x 
 + + python                         : Add optional support/bindings for the Python language
 - - python_single_target_python2_7 : Build for Python 2.7 only
 - - python_single_target_python3_4 : Build for Python 3.4 only
 - + python_single_target_python3_5 : Build for Python 3.5 only
 + - python_single_target_python3_6 : Build for Python 3.6 only
 + + python_targets_python2_7       : Build with Python 2.7
 - - python_targets_python3_4       : Build with Python 3.4
 + + python_targets_python3_5       : Build with Python 3.5
 + + python_targets_python3_6       : Build with Python 3.6
 + - roofit                         : Build RooFit function fitting package
 + - root7                          : Build experimental features to be released with ROOT 7
 + - shadow                         : Enable shadow authentication in rootd and proofd
 + + sqlite                         : Add support for sqlite - embedded sql database
 + + ssl                            : Add support for SSL/TLS connections (Secure Socket Layer / Transport Layer Security)
 + - table                          : Build libTable contrib library
 + + tbb                            : Enable implicit multi-threading (IMT) with dev-cpp/tbb
 - - test                           : Workaround to pull in packages needed to run with FEATURES=test. Portage-2.1.2 handles this internally, so don't set it
                                      in make.conf/package.use anymore
 + + threads                        : Add threads support for various packages. Usually pthreads
 + + tiff                           : Add support for the TIFF image format
 + - tmva                           : Built TMVA (ROOT's multi-variate analysis toolkit for machine learning)
 + - unuran                         : Enable support for sci-mathematics/unuran
 + - vc                             : Enable support for dev-libs/vc with VecCore
 + + xinetd                         : Add support for the xinetd super-server
 + + xml                            : Add support for XML files
 + - xrootd                         : Build the interface for net-libs/xrootd 

 * Found these USE flags for sci-physics/geant-vmc-4.3.6:
 U I
 + - doc      : Add extra documentation (API, Javadoc, etc). It is recommended to enable per package instead of globally
 + - examples : Install examples, usually source code
 + + g4root   : Build G4Root (interface for GEANT4 simulation with a ROOT geometry)
 + - geant3   : Build with Geant4 G3toG4 library
 + + mtroot   : Build MTRoot (provides ROOT IO manager classes with multi-threading support)
 + - test     : Workaround to pull in packages needed to run with FEATURES=test. Portage-2.1.2 handles this internally, so don't set it in
                make.conf/package.use anymore
 - - vgm      : Enable the Virtual Geometry Model (sci-physics/vgm)
epsft-53 ~ # 
```

amadio commented:
Nevermind, I compiled with `-O3`. Will try again with `-O2` to see if the problem goes away.

amadio commented:
I changed compile flags, but no luck. Most tests still fail for me. I'd hold on merging this until we understand why.

amadio commented:
@olifre What is the configuration you are using?

amadio commented:
@olifre I accidentally recompiled a root version from my local overlay. I will try again with the one in Gentoo later. The other big difference is that I'm using GCC 7.3.0. ROOT 6.14 is coming soon too, at the end of this month or beginning of the next, so I'd like to make sure things work with latest ROOT in any case.

olifre commented:
@amadio Of course, understood! I'd also hope it's not a problem stemming from ROOT 6.13. 
Could you share one of the logs from a failing test from:
```
 /var/tmp/portage/sci-physics/geant-vmc-4.3.6/work/geant4_vmc-3-6/examples/log
```
This should help to understand the issue.

amadio commented:
Here is one example:
```
   ------------------------------------------------------------
  | Welcome to ROOT 6.12/06                http://root.cern.ch |
  |                               (c) 1995-2017, The ROOT Team |
  | Built for linuxx8664gcc                                    |
  | From tag v6-12-06, 9 February 2018                         |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q' |
   ------------------------------------------------------------


Processing load_g4.C...
Loading Geant4 libraries (using geant4-config) ...
34
Loading g4root library ...
Loading geant4vmc library ...
Loading mtroot library ...
Error in <TReentrantRWLock::GetStateBefore()>: Must be write locked!
Processing test_E01.C("g4ConfigOld.C",kTRUE)...
Error in <TInterpreter::SnapshotMutexState>: fRecurseCount != 0 even though initial mutex state is unset!
Error in <TReentrantRWLock::GetStateBefore()>: Must be write locked!
Error in <TInterpreter::SnapshotMutexState>: fRecurseCount != 0 even though initial mutex state is unset!
Error in <TReentrantRWLock::GetStateBefore()>: Must be write locked!
Error in <TReentrantRWLock::GetStateBefore()>: Must be write locked!
Error in <TInterpreter::SnapshotMutexState>: fRecurseCount != 0 even though initial mutex state is unset!
Error in <TReentrantRWLock::GetStateBefore()>: Must be write locked!

=============================================================
 Geant4 Virtual Monte Carlo 
 Version 3.6 ( 23 January 2018 )
 WWW : http://root.cern.ch/drupal/content/geant4-vmc
=============================================================


*************************************************************
 Geant4 version Name: geant4-10-03 [MT]   (9-December-2016)
  << in Multi-threaded mode >> 
                      Copyright : Geant4 Collaboration
                      Reference : NIM A 506 (2003), 250-303
                            WWW : http://cern.ch/geant4
*************************************************************

Adding EMPhysicsList emStandard
Adding SpecialPhysicsList stepLimiter
Visualization Manager instantiating with verbosity "warnings (3)"...
Geant4 has been created.
Geometry will be defined via VMC

    TG4MCGeometry::Material:
    This method requires Geant4 installation with G3toG4.
*** TG4Exception: Aborting execution ***
```
Many fail with the same error message.

olifre commented:
@amadio Thanks! That's interesting. 
I note I have recently only tested with `sci-physics/geant[geant3]` + `sci-physics/geant-vmc[geant3]` and you appear to be using  `sci-physics/geant[geant3]` +`sci-physics/geant-vmc[-geant3]`. 

I'll check the testing code later today to see if these failures are an upstream bug in the tests (i.e. they rely on g2tog4 even if they claim not to), some hidden autodetection of g3tog4 going wrong or something completely else, and let you know.

amadio commented:
I tried disabling things to see if the tests would pass, but it also fails with geant3 enabled in geant-vmc.

mgorny commented:
@amadio, special request. Could you test and merge this when you consider it good to go? I don't really have enough free space to install Geant, and I have too much on my plate to try to find some ;-).

amadio commented:
@mgorny Sure, Geant does need lots of space for physics data. I'm already working on updating other physics software, so I will maintain this package once it hits the tree as well.

olifre commented:
> but it also fails with geant3 enabled in geant-vmc.

Are the errors the same? 

The fun fact here is that the exception you show is thrown in a variant of a function which is:
- Not even compiled in (i.e. removed by `ifdef`) when `geant3` is set. 
- Independent from `geant3` being set or unset, the function is never actually called by the tests. 

So something here is very wrong... 

To go into detail following your example output:

1. ```Geometry will be defined via VMC``` is produced at https://github.com/vmc-project/geant4_vmc/blob/1b67756148a7472067172fa97a8b1f65e802763b/examples/E01/src/Ex01MCApplication.cxx#L296 
2. Then, `ConstructMaterials` is called, which is this one: https://github.com/vmc-project/geant4_vmc/blob/1b67756148a7472067172fa97a8b1f65e802763b/examples/E01/src/Ex01DetectorConstructionOld.cxx#L49
3. Then, all `gMC->Material()` should be handled by `TVirtualMC::Material()` which in turn calls `TG4MCGeometry::Material()`. 
4. We're finally there:
https://github.com/vmc-project/geant4_vmc/blob/1b67756148a7472067172fa97a8b1f65e802763b/source/geometry/src/TG4MCGeometry.cxx#L92

Now, as you can see from https://github.com/vmc-project/geant4_vmc/blob/1b67756148a7472067172fa97a8b1f65e802763b/examples/E01/src/Ex01DetectorConstructionOld.cxx#L49 , the example *always* calls the function signature with a `Float_t*` as second-to-last argument. 

However, on your system, you end up in the function right below, which takes a `Double_t*` as second-to-last argument:
https://github.com/vmc-project/geant4_vmc/blob/1b67756148a7472067172fa97a8b1f65e802763b/source/geometry/src/TG4MCGeometry.cxx#L105
As you can see, this function (in case `USE_G3TOG4` is set, which it is if you set the useflag `geant3`) will do something. Only in case `geant3` is not set, the exception is thrown. 

So something is going very wrong on your system - it "feels" like the wrong function is called even though the signature is clearly different. 

So I have several questions:
- Any LTO on your system? 
- Is ROOT built with `-O3` by any chance? The `gMC` interface comes from ROOT, and those function calls go "through" this interface, so this might be of relevance. 
- As first step, I'd like to see the output of the same failing example when compiling with the `geant3` useflag. If the exception is still there, it can't really be a linker / function resolution problem, since the exception should not even be compiled in... 

Sorry for the long wall of text, but this looks really nasty, and I can't reproduce at all.

olifre commented:
@amadio I've just now switched one of my machines to `=sys-devel/gcc-7.3.0` (without full system recompile) and am now recompiling root, geant and then geant-vmc on it. I'll see if I can reproduce. 

But I deem it very unlikely that the preprocessor of gcc 7.3.0 is so broken that code which is ifdef'fed out is compiled in - so it's probably in any case needed to understand what goes wrong on your system. 
Could you possibly upload the full build dir of `geant-vmc` after such a broken test run (e.g on Cernbox /EOS / whatever you have access to)? Maybe I can figure out what went wrong from checking the cmake cache and all test logs.

amadio commented:
@olifre Sure, I have an update for ROOT 6.14 that I will commit soon, and I'm also working on bumping Geant4 to 10.4. I will test this again once I make these updates and post a full build log of I have problems.

olifre commented:
@amadio Perfect, thanks, also looking forward towards the updates! :+1:

olifre commented:
@amadio Do you have any idea what this (automatic?) close by @gentoogang is about?
